### PR TITLE
Add camera button to capture thumbnail at any time

### PR DESCRIPTION
**Pull request by xinan**
* Click on camera button to capture the current output and save temporarily. Then activate the `Save` button.
* Save would try to get the temp thumbnail before falling back to capture one on the spot.
* Show alert when there is an unsaved thumbnail.
* Fix problem that `unsaved changes` alert is not popping up sometimes.

### CODE REVIEW

File: content/src/controller.js [hunk 1]
```diff
@@ -131,6 +132,11 @@ function nosaveowner() {
   return model.ownername === 'frame';
 }
 
+function cansave() {
```
Comment (davidbau) [inline content/src/controller.js]: Thanks for factoring this.

File: content/src/controller.js [hunk 2]
```diff
@@ -1394,7 +1420,7 @@ function doneWithFile(filename) {
     }
   } else {
     if (filename.indexOf('/') >= 0) {
-      filename = filename.replace(/\/[^\/]+\/?$/, '');
+      filename = filename.replace(/[^\/]+\/?$/, '');
```
Comment (davidbau) [inline content/src/controller.js]: Probably right, but can you explain the change here?

File: test/edit_code.js [hunk 3]
```diff
@@ -649,6 +649,26 @@ describe('code editor', function() {
       });
     });
   });
+  it('should capture thumbnail when camera button is pressed', function(done) {
+    asyncTest(_page, one_step_timeout, null, function() {
+      // Then click the save button.
+      $('#screenshot').click();
+    }, function() {
+      // Wait for thumbnail to be flashed
+      if (!$('.tooltipster-shadow').is(':visible')) return;
```
Comment (davidbau) [inline test/edit_code.js]: Nice.  Thanks for the test.

File: content/src/controller.js (hunk 2 above)
Comment (xinan) [inline content/src/controller.js]: The reason behind this is that `history.state.previous` is something like `/edit/` or `/edit/folder/`, but we were stripping the `/` at the end, thus the comparison below fails, so the unsaved alert is not triggered. 

So the fix is to just keep the `/`.

**Comment by davidbau** [comment]
LGTM

**Pull request merged**